Optimizated calculation of shared power of 2 in bn_gcd

landgrafhomyak opened the pull request.

landgrafhomyak commented:
https://www.openssl.org/policies/cla.html:
> Every non-trivial contribution needs to be covered by a signed Contributor License Agreement (CLA) from all original authors.

It's trivial contribution, so CLA not needed.

kroeckx commented:
This looks like it undoes constant time code.

landgrafhomyak commented:
> This looks like it undoes constant time code.

Execution time of this function depends on size of input numbers, so it's not constant initially.

kroeckx commented:
This comes from https://github.com/openssl/openssl/pull/10122, everything seems to be talking about constant time. This adds a branch, making it not constant time.

Of course it depends on the size. Assuming that's public info, that's fine.

This change will probably cause a timing leak in cases where there isn't one now.

paulidale commented:
I don't think this is a trivial contribution from a copyright viewpoint.

I'm also concerned about the loss of constant time here.

Do you have any performance figures for comparison?

landgrafhomyak commented:
Updated optimization to be consttime.

I didn't do perfomance measures, but there is 64 (32 on old CPUs) times less loop iterations so it must be faster.

paulidale commented on crypto/bn/bn_gcd.c:
```diff
@@ -612,12 +612,19 @@ int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
+        mask = r->d[i] | g->d[i];
+        if ((pow2_flag == 0) | (mask == 0)) {
+            shifts += BN_BITS2 & pow2_flag;
+            continue;
```
This makes everything non-constant time.

landgrafhomyak replied:
Independent of input numbers (except `a==0 || b==0`, but this case checked before) it will run loop to check each bit in chunk exactly once:
```
mask    branch
0000    mask == 0, add 64 and continue
0000    mask == 0, add 64 and continue
0000    mask == 0, add 64 and continue
...
0000    mask == 0, add 64 and continue
1234    mask != 0, check each bit
5678    pow2_flag == 0, add 64&0 and continue
0000    pow2_flag == 0, add 64&0 and continue
...
```

kroeckx replied:
When we talk about constant time code, we're mostly talking about branch free code, where the for loop is the only branch code.

landgrafhomyak commented:
I made some measures of perfomance. New code faster ~250 times when compiled with `gcc -O0` on WSL:
```
Old | Size: 100000 Clocks per test: 27043
New | Size: 100000 Clocks per test: 1251
```
<details>
<summary>Test program (spoiler)</summary>
```cpp
#include <cstdint>
#include <cstdlib>
#include <ctime>
#include <iostream>

#define SIZE (100000)

#define TESTS (1000)

static std::uint64_t r[SIZE];
static std::uint64_t g[SIZE];

template<bool new_impl>
void test();

int main() {
    std::size_t i;
    std::srand(123);
    for (i = 0; i < SIZE; i++) {
        r[i] = (((std::uint64_t) std::rand()) << 32) | (std::uint64_t) std::rand();
        g[i] = (((std::uint64_t) std::rand()) << 32) | (std::uint64_t) std::rand();
    }

    auto start = std::clock();

    for (i = 0; i < TESTS; i++)
        test<false>();

    std::cout << "Old | Size: " << SIZE << " Clocks per test: " << (std::clock() - start) / TESTS << std::endl;

    start = std::clock();

    for (i = 0; i < TESTS; i++)
        test<true>();

    std::cout << "New | Size: " << SIZE << " Clocks per test: " << (std::clock() - start) / TESTS << std::endl;
}

template<bool new_impl>
[[clang::noinline, gnu::noinline]]
void test() {

    if constexpr (new_impl) {
        std::uint64_t pow2_temp, pow2_mask;
        int pow2_flag, pow2_chunk_index, pow2_shifts, i, j;
        pow2_flag = 1;
        pow2_chunk_index = 0;
        pow2_shifts = 0;
        for (i = 0; i < SIZE; i++) {
            pow2_mask = r[i] | g[i];
            pow2_temp = (pow2_flag != 0) & (pow2_mask != 0);
            pow2_flag &= !pow2_temp;
            pow2_temp = ((~pow2_temp & (pow2_temp - 1)) >> 63) - 1;
            pow2_shifts += 1uL & pow2_temp;
            pow2_temp = (i ^ pow2_chunk_index) & pow2_temp;
            pow2_chunk_index ^= pow2_temp;
        }
        pow2_shifts *= 64;
        pow2_mask = r[pow2_chunk_index] | g[pow2_chunk_index];
        pow2_flag = 1;
        for (j = 0; j < 64; j++) {
            pow2_flag &= pow2_mask;
            pow2_shifts += pow2_flag;
            pow2_mask >>= 1;
        }
    } else {
        std::uint64_t mask;
        int j, bit, shifts=0, i;
        bit = 1;
        for (i = 0; i < SIZE; i++) {
            mask = ~(r[i] | g[i]);
            for (j = 0; j < 64; j++) {
                bit &= mask;
                shifts += bit;
                mask >>= 1;
            }
        }
    }
}
```
</details>

t8m commented on crypto/bn/bn_gcd.c:
```diff
@@ -609,20 +609,32 @@ int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
         || !BN_lshift1(r, in_a))
         goto err;
-
+  
```
Nit: please revert this

t8m commented on crypto/bn/bn_gcd.c:
```diff
@@ -609,20 +609,32 @@ int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
-            mask >>= 1;
-        }
+        pow2_mask = r->d[i] | g->d[i];
+        pow2_temp = (pow2_flag != 0) & (pow2_mask != 0);
+        pow2_flag &= !pow2_temp;
```
I believe there is still some potential for non-const timeness here. I.e. the compiler could transfer this into branches. Could you please use routines from const_time.h appropriately?

landgrafhomyak replied:
```cpp
pow2_temp = !constant_time_is_zero(pow2_flag) & !constant_time_is_zero_64(pow2_mask);
```

t8m commented on crypto/bn/bn_gcd.c:
```diff
@@ -611,18 +612,30 @@ int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
-        }
+        pow2_mask = r->d[i] | g->d[i];
+        pow2_temp = !constant_time_is_zero(pow2_flag) & !constant_time_is_zero_64(pow2_mask);;
+        pow2_flag &= !pow2_temp;
```
pow2_flag value does not seem to be used?

landgrafhomyak replied:
It is used in previous line.

t8m commented on crypto/bn/bn_gcd.c:
```diff
@@ -611,18 +612,30 @@ int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
+        pow2_chunk_index ^= pow2_temp;
+    }
+    pow2_shifts *= BN_BITS2;
+    pow2_mask = r->d[pow2_chunk_index] | g->d[pow2_chunk_index];
```
This dereference won't be constant time either because here you're indexing based on the values of the numbers.

You need to calculate the pow2_mask value within the for loop above.

kroeckx commented:
We have a lot of test failing in CI, so I assume the result of the calculation is wrong now.

t8m commented:
I still have some suspicions about how the compiler might make the logical `!` operations non-const time. Perhaps the const_time_select...() functions could be used instead? Otherwise this looks fairly good to me.

kroeckx commented:
If this is merged, we might want to reevaluate if ossl_bn_rsa_fips186_4_derive_prime() should use gcd or not.

paulidale commented:
> I made some measures of perfomance. New code faster ~25 times when compiled with `gcc -O0` on WSL:
> 
> ```
> Old | Size: 100000 Clocks per test: 27043
> New | Size: 100000 Clocks per test: 1251
> ```

I think it would be enlightening to see the min and max times.  That would confirm constant time execution.

romen commented:
I will try to find the time for a proper review, but I am a bit swamped at the moment.

The first thing that came to mind inspecting the code changes, is that as a general rule of thumb, one generally wants to avoid mixing bitwise operations and the `! ` operator.
Quite often the `! ` is processed by the compiler in a way that introduces a branch in the resulting binary code, or that could be handled as a branch by the hw optimization heuristics.
I computing this kind of bitwise masks, we tend to prefer using only bitwise operators `&¦^~` as that usually prevents optimization steps to unfold the computation into branches.

I might try to summon @cpereida and @bbbrumley if they can volunteer their expertise.

kroeckx commented:
@landgrafhomyak Will you provide us with a CLA?

landgrafhomyak commented:
> I think it would be enlightening to see the min and max times. That would confirm constant time execution.

I tested on general-purpose computer, so time of execution depends on a lot of factors:
```
Fixed old | Clocks: 151596 <= 213936 <= 335924
Fixed new | Clocks: 6376 <= 7781.11 <= 10398
Random old | Clocks: 156716 <= 200857 <= 311670
Random new | Clocks: 6667 <= 10194.9 <= 15905
```
As you can see time differs up to 2 times for same test inputs.

<details>
  <summary>Test program (spoiler)</summary>

  ```cpp
  #include <cstdint>
  #include <cstdlib>
  #include <ctime>
  #include <string>
  #include <iostream>

  #define SIZE (1000000)

  #define TESTS (100)

  volatile static std::uint64_t R[SIZE];
  volatile static std::uint64_t G[SIZE];

  template<bool new_impl>
  void test(std::uint64_t *r, std::uint64_t *g);

  void fill(std::uint64_t *r, std::uint64_t *g);

  void measure(std::string name, bool regenerate, void (fn)(std::uint64_t *, std::uint64_t *));

  int main() {

    std::srand(123);

    fill((std::uint64_t *) R, (std::uint64_t *) G);

    measure("Fixed old", false, test<false>);
    measure("Fixed new", false, test<true>);
    measure("Random old", true, test<false>);
    measure("Random new", true, test<true>);
  }


  void measure(std::string name, bool regenerate, void (fn)(std::uint64_t *, std::uint64_t *)) {

    int i;
    unsigned long long min = -1, max = 0, sum = 0;
    for (i = 0; i < TESTS; i++){
        if (regenerate)
            fill((std::uint64_t *) R, (std::uint64_t *) G);
        auto start = std::clock();
        fn((std::uint64_t *) R, (std::uint64_t *) G);
        auto delta = std::clock() - start;
        if (delta < min) min = delta;
        if (delta > max) max = delta;
        sum += delta;
    }
    std::cout << name << " | Clocks: " << min << " <= " << (((double) sum) / TESTS) << " <= " << max << std::endl;
  }

  void fill(std::uint64_t *r, std::uint64_t *g) {
    std::size_t i;
    for (i = 0; i < SIZE; i++) {
        r[i] = (((std::uint64_t) std::rand()) << 32) | (std::uint64_t) std::rand();
        g[i] = (((std::uint64_t) std::rand()) << 32) | (std::uint64_t) std::rand();
    }
  }

  namespace const_time {
    std::uint64_t msb(std::uint64_t v) {
        return 0 - (v >> 63);
    }

    std::uint64_t is_zero(std::uint64_t v) {
        return msb(~v & (v - 1));
    }

    std::uint64_t barrier(std::uint64_t v) {
        volatile std::uint64_t u = v;
        return u;
    }

    std::uint64_t select(std::uint64_t m, std::uint64_t t, std::uint64_t f) {
        return (barrier(m) & t) | (barrier(~m) & f);
    }
  }

  template<bool new_impl>
  void test(std::uint64_t *r, std::uint64_t *g) {
    if constexpr (new_impl) {
        std::uint64_t pow2_temp, pow2_mask;
        int pow2_flag, pow2_chunk_index, pow2_shifts, i, j;
        pow2_flag = 1;
        pow2_chunk_index = 0;
        pow2_shifts = 0;
        for (i = 0; i < SIZE; i++) {
            pow2_mask = r[i] | g[i];
            pow2_temp = (pow2_flag != 0) & (pow2_mask != 0);
            pow2_flag &= !pow2_temp;
            pow2_temp = ((~pow2_temp & (pow2_temp - 1)) >> 63) - 1;
            pow2_shifts += 1uL & pow2_temp;
            pow2_temp = (i ^ pow2_chunk_index) & pow2_temp;
            pow2_chunk_index ^= pow2_temp;
        }
        pow2_shifts *= 64;
        pow2_mask = r[pow2_chunk_index] | g[pow2_chunk_index];
        pow2_flag = 1;
        for (j = 0; j < 64; j++) {
            pow2_flag &= pow2_mask;
            pow2_shifts += pow2_flag;
            pow2_mask >>= 1;
        }
    } else {
        std::uint64_t mask;
        int j, bit, shifts = 0, i;
        bit = 1;
        for (i = 0; i < SIZE; i++) {
            mask = ~(r[i] | g[i]);
            for (j = 0; j < 64; j++) {
                bit &= mask;
                shifts += bit;
                mask >>= 1;
            }
        }
    }
  }


  ```
</details>

bbbrumley commented:
@landgrafhomyak so the idea is to first identify the number of shared LSB zero words in constant time, multiply that by the word size, then work on the sub-word zero bits only on that specific word, instead of every word along the way?

landgrafhomyak commented:
> Will you provide us with a CLA?

Sent.

> idea is to first identify the number of shared LSB zero words in constant time, multiply that by the word size, then work on the sub-word zero bits only on that specific word, instead of every word along the way?

Yes.

<hr>

In CI of last commit one check [failed](https://github.com/openssl/openssl/actions/runs/9000361153/job/24729221412?pr=24332) because of fetch issues:
```
fatal: unable to access 'https://github.com/ngtcp2/nghttp3.git/': Failed to connect to github.com port 443 after 133642 ms: Connection timed out
```

bbbrumley commented:
> > idea is to first identify the number of shared LSB zero words in constant time, multiply that by the word size, then work on the sub-word zero bits only on that specific word, instead of every word along the way?
> 
> Yes.

OK yea, that seems clever :)

Given the nature of the change, I would suggest adding test cases [here](https://github.com/openssl/openssl/blob/master/test/recipes/10-test_bn_data/bngcd.txt), such that one of the middle words of one and both arguments are zero, aligned and unaligned. Like

```
A = deadbeefdeadbeef0000000000000000cafebabecafebabe
B = 12345678123456780000000000000000abcdabcdabcdabcd
```

Then maybe sth like

```
A = 432143214321432100000000000000001234123412341234abcdabcdabcdabcd
B = 6789678967896789987698769876987600000000000000004321432143214321
```

The "LSB word" test cases are already well-represented in the corpus.

bbbrumley commented:
Thank you for adding the unit tests! I think this is a solid contribution.

t8m commented:
Commits to be squashed when merging.

landgrafhomyak commented on crypto/bn/bn_gcd.c:
```diff
@@ -611,18 +612,28 @@ int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
-        }
+        pow2_numbits_temp = r->d[i] | g->d[i];
+        pow2_condition_mask = ~(constant_time_is_zero_bn(pow2_flag)) & ~(constant_time_is_zero_bn(pow2_numbits_temp));
+        pow2_flag &= ~pow2_condition_mask;
```
I prefer keep code clear, there is two conditions:
* flag != 0
* mask != 0

Compiler is clever enough to optimize this.

bernd-edlinger replied:
I am not concerned about compiler.  Just trying to understand
code with triple negation is a bit hard for humans.
but how about
```C
 pow2_condition_mask = constant_time_is_zero_bn(pow2_numbits_temp);
 pow2_flag &= pow2_condition_mask;
```
wouldn't that also work, because pow2_flag is initialized with 1 and can only have one falling edge?

landgrafhomyak replied:
> I am not concerned about compiler. Just trying to understand code with triple negation is a bit hard for humans.

Human should understand that `pow2_condition_mask` has `true` (`0xF...F` in this case) if condition to update met.

Updating flag is less understandable, but it's cost of consttime calculation.


> ```c
>  pow2_condition_mask = constant_time_is_zero_bn(pow2_numbits_temp);
>  pow2_flag &= pow2_condition_mask;
> ```
> 
> wouldn't that also work, because pow2_flag is initialized with 1 and can only have one falling edge?

This seems broke subsequent uses of `pow2_condition_mask`: it will be in *condition met* state more than once.

landgrafhomyak replied:
> Hmm, now I see a way how to optimize the code even a bit more:
> 
> ```c
>     for (i = 0; i < r->dmax && i < g->dmax; i++) {
>         pow2_numbits_temp = r->d[i] | g->d[i];
>         pow2_condition_mask = constant_time_is_zero_bn(pow2_flag);
>         pow2_flag &= constant_time_is_zero_bn(pow2_numbits_temp);
>         pow2_shifts += pow2_flag;
>         pow2_numbits = constant_time_select_bn(pow2_condition_mask,
>                                                pow2_numbits, pow2_numbits_temp);
>     }
> ```
> 
> how about that?

Okay, this seems good.



> > I prefer keep code clear, there is two conditions: flag != 0 && mask != 0, not flag == 0 || mask == 0
> 
> Were those conditions in previous versions of this patch? When I look at the current version of the patch I don't really see them.

This expression was here since 3rd commit (when i removed `if(...) continue`), initially with `!= 0`, than `~constant_time_is_zero_bn`

t8m commented on crypto/bn/bn_gcd.c:
```diff
@@ -611,18 +612,28 @@ int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
+        pow2_numbits_temp = r->d[i] | g->d[i];
+        pow2_condition_mask = ~(constant_time_is_zero_bn(pow2_flag)) & ~(constant_time_is_zero_bn(pow2_numbits_temp));
+        pow2_flag &= ~pow2_condition_mask;
+        pow2_shifts += 1 & pow2_flag;
```
Is this `1 &` operation redundant? As pow2_flag is starting at 1 and only `&=` is used with it.

bernd-edlinger commented on crypto/bn/bn_gcd.c:
```diff
@@ -611,18 +612,28 @@ int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
+    pow2_shifts *= BN_BITS2;
+    pow2_flag = 1;
+    for (j = 0; j < BN_BITS2; j++) {
+        pow2_flag &= pow2_numbits; 
```
There is a space at end of line here, please remove it

bernd-edlinger commented on crypto/bn/bn_gcd.c:
```diff
@@ -580,8 +581,8 @@ int BN_are_coprime(BIGNUM *a, const BIGNUM *b, BN_CTX *ctx)
-    BN_ULONG mask = 0;
-    int i, j, top, rlen, glen, m, bit = 1, delta = 1, cond = 0, shifts = 0, ret = 0;
+    BN_ULONG pow2_numbits, pow2_numbits_temp, pow2_condition_mask;
+    int i, j, top, rlen, glen, m, delta = 1, cond = 0, pow2_shifts, ret = 0, pow2_flag;
```
this line is longer than 80 lines, maybe move the pow2_flag to the line above, making it a
BN_ULONG, that would avoid implicit type casts in `constant_time_is_zero_bn(pow2_flag)` below

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Squashed and merged to the master branch. Thank you for your contribution.
